src/Microsoft.ML.Data/DataView/TypedCursor.cs

```diff
@@ -217,6 +217,90 @@ public RowCursor<TRow>[] GetCursorSet(Func<int, bool> additionalColumnsPredicate
+                    InternalSchemaDefinition.GetMappedType(propertyInfo.PropertyType, out actualType, out isVector);
+                    customAttributes = propertyInfo.GetCustomAttributes();
+                    break;
+
+                default:
+                    Contracts.Assert(false);
+                    throw Contracts.ExceptNotSupp("Expected a FieldInfo or a PropertyInfo");
+            }
+
+            if (!actualType.TryGetDataKind(out _) && !DataViewTypeManager.Knows(actualType, customAttributes))
+            {
+                int colIndex;
+                data.Schema.TryGetColumnIndex(singleName, out colIndex);
+                DataViewType expectedType = data.Schema[colIndex].Type;
+                if (!actualType.Equals(expectedType.RawType))
+                    throw Contracts.ExceptParam(nameof(actualType), $"The expected type '{expectedType.RawType}' does not match the type of the '{singleName}' property: '{actualType.Name}'. Please change the {singleName} property to '{expectedType.RawType}'");
+            }
+        }
+
+        public static void ValidateUserType(SchemaDefinition schemaDefinition, Type userType, IDataView data)
+        {
+            //Get memberInfos
+            MemberInfo[] memberInfos = null;
+            if (schemaDefinition == null)
+            {
+                memberInfos = SchemaDefinition.GetMemberInfos(userType, SchemaDefinition.Direction.Write);
+            }
+            else
+            {
+                for (int i = 0; i < schemaDefinition.Count; ++i)
+                {
+                    var col = schemaDefinition[i];
+                    if (col.MemberName == null)
+                        throw Contracts.ExceptParam(nameof(schemaDefinition), "Null field name detected in schema definition");
+
+                    MemberInfo memberInfo = null;
+                    // Infer the column name.
+                    var colName = string.IsNullOrEmpty(col.ColumnName) ? col.MemberName : col.ColumnName;
+
+                    if (col.Generator == null)
+                    {
+                        memberInfo = userType.GetField(col.MemberName);
+
+                        if (memberInfo == null)
+                            memberInfo = userType.GetProperty(col.MemberName);
+
+                        if ((memberInfo is FieldInfo && (memberInfo as FieldInfo).FieldType == typeof(IChannel)) ||
+                        (memberInfo is PropertyInfo && (memberInfo as PropertyInfo).PropertyType == typeof(IChannel)))
+                            continue;
+                    }
+                    else
+                    {
+                        memberInfo = col.ReturnType;
+                    }
+                    ValidateMemberInfo(memberInfo, data);
+                }
+            }
+
+            if (memberInfos == null)
+                return;
+
+            foreach (var memberInfo in memberInfos)
+                ValidateMemberInfo(memberInfo, data);
```

Comment:
I'm a bit confused by this. It looks to me that the `memberInfos` variable is only used inside the case were `schemaDefinition == null`, so I think these lines can be moved inside the `schemaDefinition == null` branch.